### PR TITLE
add timeout to dial

### PR DESCRIPTION
**Pull request by ducquangkstn**
- Add a function to try to connect with timeout

### CODE REVIEW

File: serviceuser.go [hunk 1]
```diff
@@ -160,6 +161,23 @@ func (su *ServiceUser) waitUntilReady() error {
 	return nil
 }
 
+// Connect connects to the server at the given "host:port". Either Connect or
```
Comment (favadi) [inline serviceuser.go]: Comment should start with ConnectWithTimeout.

File: serviceuser.go [hunk 2]
```diff
@@ -160,6 +161,23 @@ func (su *ServiceUser) waitUntilReady() error {
...
+// SetConn must be before calling CStore, etc.
+func (su *ServiceUser) ConnectWithTimeout(serverAddr string, timeout time.Duration) error {
```
Comment (favadi) [inline serviceuser.go]: I think using a context here might be useful later, if we want to graceful terminate the Connect process. I image it should look like

```go
func (su *ServiceUser) ConnectContext(ctx context.Context, serverAddr string) {
	var dialer net.Dialer
	conn, err := dialer.DialContext(ctx, "tcp", serverAddr) 
}

ctx, cancel := context.WithTimeout(time.Minute)
su.ConnectWithTimeout(ctx, ...)
```

File: serviceuser.go [hunk 3]
```diff
@@ -160,6 +161,23 @@ func (su *ServiceUser) waitUntilReady() error {
...
+		return fmt.Errorf("net.DialTimeout: %w", err)
+	} else {
+		su.disp.downcallCh <- stateEvent{event: evt02, pdu: nil, err: nil, conn: conn}
+	}
+	return nil
+}
+
 // Connect connects to the server at the given "host:port". Either Connect or
 // SetConn must be before calling CStore, etc.
 func (su *ServiceUser) Connect(serverAddr string) {
```
Comment (favadi) [inline serviceuser.go]: The Connect and ConnectWithTimeout are mostly duplicated. We could turn this old function into

```
func (su *ServiceUser) Connect(serverAddr string) {
    return su.ConnectContext(context.Background(), serverAddr)
}
```

**Review by favadi — CHANGES_REQUESTED** [review]

File: serviceuser.go (hunk 1 above)
Comment (ducquangkstn) [reply serviceuser.go]: Fixed at f6f6d07704b9323da4735839ed0f1eee60b0f1f6

File: serviceuser.go (hunk 2 above)
Comment (ducquangkstn) [reply serviceuser.go]: Fixed at f6f6d07704b9323da4735839ed0f1eee60b0f1f6

File: serviceuser.go (hunk 3 above)
Comment (ducquangkstn) [reply serviceuser.go]: Fixed at f6f6d07704b9323da4735839ed0f1eee60b0f1f6

**Review by favadi — APPROVED** [review]

**Pull request merged**